src/Analysis/Ast/Impl/Analyzer/PythonAnalyzerSession.cs

```diff
@@ -230,136 +253,314 @@ internal sealed class PythonAnalyzerSession {
+                    _log?.Log(TraceEventType.Verbose, $"Attempt to reanalyze fully cached modules cycle: {names}");
+                }
+            }
+
+            var imports = new List<(AnalysisModuleKey From, int FromPosition, AnalysisModuleKey To, int ToPosition)>();
+            foreach (var (fromModule, fromPosition, toModule, toName) in importNames) {
+                if (!entries.ContainsKey(toModule)) {
+                    continue;
                 }
-                if (ast == null) {
-                    if (currentAnalysis == null) {
-                        // Entry doesn't have ast yet. There should be at least one more session.
-                        Cancel();
-                        _log?.Log(TraceEventType.Verbose, $"Analysis of {module.Name}({module.ModuleType}) canceled (no AST yet).");
-                        return false;
+
+                if (toName == null) {
+                    imports.Add((fromModule, fromPosition, toModule, 0));
+                } else if (variables.TryGetValue((toModule, toName), out var toPosition)) {
+                    imports.Add((fromModule, fromPosition, toModule, toPosition));
+                }
+            }
+            
+            var startingKeys = LocationLoopResolver<AnalysisModuleKey>.FindStartingItems(imports);
+            lock (_syncObj) {
+                if (_isCanceled) {
+                    return;
+                }
+            }
+
+            var variableHandler = new LoopImportedVariableHandler(_services, asts, cachedVariables, IsCanceled);
+            foreach (var key in startingKeys) {
+                if (asts.TryGetValue(key, out var startingAst) && entries.TryGetValue(key, out var me)) {
+                    variableHandler.WalkModule(me.Module, startingAst);
+                }
+
+                lock (_syncObj) {
+                    if (_isCanceled) {
+                        return;
                     }
-                    //Debug.Fail($"Library module {module.Name} of type {module.ModuleType} has been analyzed already!");
-                    return false;
                 }
+            }
+
+            foreach (var walker in variableHandler.Walkers) {
+                asts.Remove(new AnalysisModuleKey(walker.Module));
+            }
+
+            while (asts.Count > 0) {
+                var (moduleKey, ast) = asts.First();
+                variableHandler.WalkModule(entries[moduleKey].Module, ast);
+                
+                lock (_syncObj) {
+                    if (_isCanceled) {
+                        return;
+                    }
+                }
+
+                foreach (var walker in variableHandler.Walkers) {
+                    asts.Remove(new AnalysisModuleKey(walker.Module));
+                }
+            }
+
+            foreach (var walker in variableHandler.Walkers) {
+                var module = (IDocument)walker.Module;
+                var entry = entries[new AnalysisModuleKey(module)].Entry;
+
```

Comment:
`ActivityTracker.OnEnqueueModule(entry.Module.FilePath)` here will make the tracer accurate again.